luarules/gadgets/unit_weapon_smart_select_helper.lua

```diff
@@ -0,0 +1,217 @@
+		return true
+	else
+		return false
+	end
+end
+
+
+local function updateAimingState(attackerID)
+	local data = smartUnits[attackerID]
+	local defData = unitDefData[data.unitDefID]
+
+	local priorityTargetType, priorityIsUserTarget, priorityTarget = spGetUnitWeaponTarget(attackerID, defData.priorityWeapon)
+	local backupIsUserTarget, backupTarget = select(2, spGetUnitWeaponTarget(attackerID, defData.backupWeapon))
+
+	local preferredCanShoot = false
+	if priorityTargetType == PRIORITY_AIMINGSTATE then
+		preferredCanShoot = spGetUnitWeaponHaveFreeLineOfFire(attackerID, defData.trajectoryCheckWeapon, priorityTarget)
+	elseif priorityTargetType == BACKUP_AIMINGSTATE then
+		preferredCanShoot = spGetUnitWeaponHaveFreeLineOfFire(attackerID, defData.trajectoryCheckWeapon, nil, nil, nil,
+			priorityTarget[1], priorityTarget[2], priorityTarget[3])
+	end
+
+	if not data.suspendErrorUntilFrame and (backupTarget or priorityTarget) then
+		data.suspendErrorUntilFrame = gameFrame + defData.failedToFireFrameThreshold
+	elseif not backupTarget and not priorityTarget then
+		data.suspendErrorUntilFrame = nil
+	end
+
+	local failureToFire = false
+	if defData.canMove then
+		if not spGetUnitEstimatedPath(attackerID) then --if it's moving, it'll probably fail to shoot from slow turn speed, or gain LOS from movement
+			failureToFire = failureToFireCheck(attackerID, data, defData)
+		end
+	else
+		failureToFire = failureToFireCheck(attackerID, data, defData)
+	end
+
+	if priorityIsUserTarget and preferredCanShoot then
+		if failureToFire then
+			data.errorTallyMultiplier = data.errorTallyMultiplier + errorMultiplierAddition
+			data.aggroBias = backupErrorAggro * data.errorTallyMultiplier ^ data.errorTallyMultiplier
+		else
+			data.aggroBias = data.aggroBias + priorityManualAggro
+		end
+	elseif backupIsUserTarget then
+		data.aggroBias = data.aggroBias - backupManualAggro
+	else
+		if failureToFire then
+			data.errorTallyMultiplier = data.errorTallyMultiplier + errorMultiplierAddition
+			data.aggroBias = backupErrorAggro * data.errorTallyMultiplier ^ data.errorTallyMultiplier
+			data.suspendErrorUntilFrame = gameFrame + defData.failedToFireFrameThreshold
+		elseif preferredCanShoot then
+			data.aggroBias = data.aggroBias + priorityAutoAggro
+		elseif backupIsUserTarget ~= nil then
+			data.aggroBias = data.aggroBias - backupAutoAggro
+		end
+	end
+
+	data.errorTallyMultiplier = data.errorTallyMultiplier * errorTallyDecayRate
+	if data.aggroBias >= prioritySwitchThreshold then
+		data.aggroBias = mathMax(data.aggroBias * aggroDecayRate, data.aggroBias - aggroDecayCap)
+		spCallCOBScript(attackerID, data.setStateScriptID, 0, PRIORITY_AIMINGSTATE)
+	elseif data.aggroBias < backupSwitchThreshold then
+		data.aggroBias = mathMin(data.aggroBias * aggroDecayRate, data.aggroBias + aggroDecayCap)
+		spCallCOBScript(attackerID, data.setStateScriptID, 0, BACKUP_AIMINGSTATE)
+	else
+		data.aggroBias = data.aggroBias * aggroDecayRate
+	end
```

Comment:
Is all this complex math needed? Can this be accomplished by just keeping a tally of the number of failed aiming attempts?

Reply:
- each subsequent failure must be compoundingly more punishing then the previous 
- The error state must decay, because the conditions that caused it to have an error may be transient  (like being built behind dragon's teeth)
- simple number tally did not produce desirable results. Multipliers took its place to allow me greater flexibility in tuning behavior